# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [42]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [24]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
lending_data_df = pd.read_csv(Path('/content/drive/MyDrive/Challenge12/Resources/lending_data.csv'))

# Review the DataFrame
# YOUR CODE HERE!
lending_data_df.head(-5)

loan_size  interest_rate  ...  total_debt  loan_status
0        10700.0          7.672  ...       22800            0
1         8400.0          6.692  ...       13600            0
2         9000.0          6.963  ...       16100            0
3        10700.0          7.664  ...       22700            0
4        10800.0          7.698  ...       23000            0
...          ...            ...  ...         ...          ...
77526    18300.0         10.895  ...       53100            1
77527    20900.0         11.988  ...       63400            1
77528    15100.0          9.557  ...       40500            1
77529    19300.0         11.347  ...       57400            1
77530    19700.0         11.508  ...       58900            1

[77531 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [8]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = lending_data_df['loan_status']
# Separate the X variable, the features
# YOUR CODE HERE!
X = lending_data_df.drop(columns='loan_status')

In [9]:
# Review the y variable Series
# YOUR CODE HERE!
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [10]:
# Review the X variable DataFrame
# YOUR CODE HERE!
X[:5]

loan_size  interest_rate  ...  derogatory_marks  total_debt
0    10700.0          7.672  ...                 1       22800
1     8400.0          6.692  ...                 0       13600
2     9000.0          6.963  ...                 0       16100
3    10700.0          7.664  ...                 1       22700
4    10800.0          7.698  ...                 1       23000

[5 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [11]:
# Check the balance of our target values
# YOUR CODE HERE!
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [12]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [25]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
logistic_regression_model = LogisticRegression(random_state = 1)

# Fit the model using training data
# YOUR CODE HERE!
logistic_regression_model.fit(X_train, y_train)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=1, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [27]:
# Make a prediction using the testing data
# YOUR CODE HERE!
predictions = lending_model.predict(X_train)
training_predictions = logistic_regression_model.predict(X_train)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [28]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
print(balanced_accuracy_score(y_train, training_predictions))


0.9494259906569136


In [29]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_train, training_predictions)


array([[55994,   277],
       [  181,  1700]])

In [30]:
# Print the classification report for the model
# YOUR CODE HERE!
print(classification_report_imbalanced(y_train, training_predictions))


                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      1.00      0.90      1.00      0.95      0.91     56271
          1       0.86      0.90      1.00      0.88      0.95      0.89      1881

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     58152



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**  healthy loans correct 100% of the time, high risk loans 85% of the time.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [31]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
random_oversampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
X_resampled, y_resampled = random_oversampler.fit_resample(X_train, y_train)


In [43]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
y_resampled.value_counts()


AttributeError: ignored

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [38]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
logistic_regression_model = LogisticRegression(random_state = 1)

# Fit the model using the resampled training data
# YOUR CODE HERE!
logistic_regression_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
# YOUR CODE HERE!
testing_predictions = logistic_regression_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [39]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
print(balanced_accuracy_score(y_test, testing_predictions))


0.9936781215845847


In [40]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
print(confusion_matrix(y_test, testing_predictions))


[[18649   116]
 [    4   615]]


In [44]:
# Print the classification report for the model
# YOUR CODE HERE!
print(classification_report(y_test, testing_predictions))


              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** YOUR ANSWER HERE!
The resampled model identified the 1 Class with only 84% pre and 99% rec. The model incorrectly classified 10-15% of the 1 Class predictions

The Resampled Model identified the 0 Class with 100% pre and 99% rec. The model correctly classified effectively 100% of the 0 Class predictions

